c2rust-transpile/src/translator/mod.rs

```diff
@@ -2231,6 +2231,22 @@ impl<'c> Translation<'c> {
                     .unwrap_or((*id, ty));
                 let expr = self.convert_expr(ctx, expr_id)?;
 
+                // Issue: https://github.com/immunant/c2rust/issues/803
+                // This actually should be catched by `to_unsafe_pure_expr` see below
+                // but a function call expression is translated as having no sideeffects by `convert_expr` (above)
+                // which is not necessarily correct, we just might not know them (e.g only decl in header exists).
+                // So we catch this here until there is a way to express unknown sideeffects of an expression.
+                let expr = match expr.result_map(|v| {
+                    if matches!(*v, Expr::Call(_)) {
+                        Err(())
+                    } else {
+                        Ok(v)
+                    }
+                }) {
+                    Ok(v) => v,
+                    Err(_) => return Err(format_err!("Can not expand macro to function call")),
+                };
```

Comment:
This can be written more simply:

```suggestion
                let expr = expr.result_map(|v| {
                    if matches!(*v, Expr::Call(_)) {
                        Err(format_err!("Macro definition contains function call"))
                    } else {
                        Ok(v)
                    }
                })?;
```